tuning/data/data_processors.py

```diff
@@ -0,0 +1,238 @@
+# Copyright The FMS HF Tuning Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Standard
+from abc import ABC, abstractmethod
+from typing import Dict, List, Union
+import logging
+import os
+
+# Third Party
+from datasets import Dataset, DatasetDict, IterableDataset
+from datasets.exceptions import DatasetNotFoundError
+from transformers import AutoTokenizer
+import datasets
+import torch
+
+# Local
+from tuning.data.data_config import DataConfig, DataLoaderConfig, DataSetConfig
+from tuning.data.data_handlers import AVAILABLE_DATA_HANDLERS
+from tuning.utils.utils import get_extension, get_loader_for_filepath
+
+
+class DataPreProcessor(ABC):
+
+    tokenizer = None
+    data_config: DataConfig = None
+    dataloaderconfig: DataLoaderConfig = None
+    registered_handlers: Dict[str, callable] = None
+
+    def __init__(self, dataloaderconfig: DataLoaderConfig, tokenizer: AutoTokenizer):
+        self.tokenizer = tokenizer
+        self.dataloaderconfig = dataloaderconfig
+
+        # Initialize other objects
+        self.registered_handlers = {}
+
+    def load_dataset(
+        self,
+        datasetconfig: DataSetConfig,
+        splitName: str,
+        datafile: str = None,
+        **kwargs,
+    ):
+        raise NotImplementedError("Needs to be implemented")
+
+    def register_data_handler(self, name: str, func: callable):
+        self.registered_handlers[name] = func
+
+    @abstractmethod
+    def process_dataset_configs(
+        self, dataset_configs: List[DataSetConfig], **extra_kwargs
+    ) -> Union[Dataset, IterableDataset]:
+        raise NotImplementedError("Needs to be implemented")
+
+
+class HFBasedDataPreProcessor(DataPreProcessor):
```

Comment:
it is not clear, besides this one implemebntation, what other implementations we will need

Reply:
This one's a bit tricky because initially we wanted to keep the implementation modular to make sure this can be extended to any other requirements later (think maybe a data preprocessor built on top of custom torch datasets) but since we do not have any active requirements in the roadmap and such experimental features might not be directly integrated, I am dropping the abstract class in favour of just HF implementation.